qutip/lattice.py

```diff
@@ -0,0 +1,785 @@
+                        data = np.append(data,[np.conj(inter_hop_y[l,k]),  inter_hop_y[l,k] ]);
+
+    for j in range(0,ny_units):
+        lin_RI = j;
+        if (PBCx == 1 and nx_units*cell_num_site > 2):
+            for k in range(x0):
+                for l in range(y0):
+                    (Aw,)=np.where(row_ind == (lin_RI*nx_units*NS+k)  )
+                    (Bw,)=np.where(col_ind == (((lin_RI+1)*nx_units-1)*NS+l)   )                    
+                    if ( len(np.intersect1d(Aw,Bw)  ) == 0 ):                    
+                        for m in range(xi_len):
+                            row_ind = np.append(row_ind,[lin_RI*nx_units*NS+k,  ((lin_RI+1)*nx_units-1)*NS+l ]);
+                            col_ind = np.append(col_ind,[((lin_RI+1)*nx_units-1)*NS+l,  lin_RI*nx_units*NS+k]);
+                            data = np.append(data,[np.conj(inter_hop_x[m][l,k]), inter_hop_x[m][l,k] ]);
+            
+    Hamt = csr_matrix((data, (row_ind, col_ind)), [M, M], dtype=np.complex )            
+    return Hamt
+
+def diag_a_matrix( H_k, calc_evecs = False):
+    """
+    Returns eigen-values and/or eigen-vectors of an input matrix.
+    
+    Parameters
+    ==========
+    H_k : numpy matrix
+        The matrix to be diagonalized
+            
+    Returns
+    -------
+    vecs : numpy ndarray
+        vecs[:,:] = [band index,:] = eigen-vector of band_index
+    vals : numpy ndarray
+        The diagonalized matrix                        
+    """                 
+    if np.max(H_k-H_k.T.conj())>1.0E-20:
+        raise Exception("\n\nThe Hamiltonian matrix is not hermitian!")
+
+    if calc_evecs == False: # calculate only the eigenvalues
+        vals=np.linalg.eigvalsh(H_k.todense())
+        return np.array(vals,dtype=float)
+    else: # find eigenvalues and eigenvectors
+        (vals, vecs)=np.linalg.eigh(H_k.todense())
+        vecs=vecs.T    
+        # now vecs[i,:] is eigenvector for vals[i,i]-th eigenvalue        
+        return (vals,vecs)
+
+def cell_structures( val_s = [], val_t = [], val_u = [], val_v = [], limit = []):
+    """
+    Returns two matrices cell_H and cell_T to help the user form the inputs for
+    defining an instance of Lattice1d and Lattice2d classes. The two matrices
+    are the intra and inter cell Hamiltonians with the tensor structure of the 
+    specified site numbers and/or degrees of freedom defined by the user.
+    
+    Parameters
+    ==========
+    val_s : list of str
+        The first list of str's specifying the sites/degrees of freedom in the
+        unitcell
+        
+    val_t : list of str
+        The second list of str's specifying the sites/degrees of freedom in the
+        unitcell
+        
+    val_u : list of str
+        The third list of str's specifying the sites/degrees of freedom in the
+        unitcell
+        
+    val_v : list of str
+        The fourth list of str's specifying the sites/degrees of freedom in the
+        unitcell
```

Comment:
Four input with almost the same description, very confusing.
The last input is a dummy input that if used, print but not raise and error... Please get rid of it.